Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -352,4 +372,337 @@ namespace MaterialCanvas
+            }
+            if (auto v = AZStd::any_cast<const AZ::Vector2>(&slotValue))
+            {
+                return AZStd::string::format("float2(%g, %g)", v->GetX(), v->GetY());
+            }
+            if (auto v = AZStd::any_cast<const float>(&slotValue))
+            {
+                return AZStd::string::format("%g", *v);
+            }
+            if (auto v = AZStd::any_cast<const int>(&slotValue))
+            {
+                return AZStd::string::format("%i", *v);
+            }
+            if (auto v = AZStd::any_cast<const unsigned int>(&slotValue))
+            {
+                return AZStd::string::format("%u", *v);
+            }
+            return AZStd::string();
+        };
+
+        // Gather all of the common settings usually associated with a material canvas note to hand slot.
+        auto collectCommonSettings = [&](const AtomToolsFramework::DynamicNodeSettingsMap& settings)
+        {
+            collectSettingsAsSet(settings, "includePaths", includePaths);
+            collectSettingsAsSet(settings, "templatePaths", templatePaths);
+            collectSettingsAsVec(settings, "classDefinitions", classDefinitions);
+            collectSettingsAsVec(settings, "functionDefinitions", functionDefinitions);
+        };
+
+        // Each instruction block needs to have name substitutions performed on it. Variable names need to be unique per
+        // node and will be prepended with the node ID. Slot types and values will be determined based on connections and
+        // then substituted.
+        auto updateAndAddSlotInstructions =
+            [&](GraphModel::ConstNodePtr node, GraphModel::ConstSlotPtr slot, AZStd::vector<AZStd::string>& instructionsForSlot)
+        {
+            replaceStringsInVec("NODEID", AZStd::string::format("node%u", node->GetId()), instructionsForSlot);
+            replaceStringsInVec("SLOTNAME", slot->GetName().c_str(), instructionsForSlot);
+            replaceStringsInVec("SLOTTYPE", convertSlotTypeName(slot->GetDataType()->GetDisplayName()), instructionsForSlot);
+            replaceStringsInVec("SLOTVALUE", convertSlotValue(slot->GetValue()), instructionsForSlot);
+            instructions.insert(instructions.end(), instructionsForSlot.begin(), instructionsForSlot.end());
+        };
+
+        AZStd::vector<GraphModel::ConstNodePtr> nodes = GetNodesInExecutionOrder();
+
+        for (const auto& node : nodes)
+        {
+            if (auto dynamicNode = azrtti_cast<const AtomToolsFramework::DynamicNode*>(node.get()))
+            {
+                const auto& nodeConfig = dynamicNode->GetConfig();
+
+                // Prescan to see if this node has any output connections
+                // Generally speaking, we can ignore the node if there are no output connections but we're just trying to gather data now.
+                bool hasOutputConnections = false;
+                for (const auto& slotConfig : nodeConfig.m_outputSlots)
```

Comment:
So if I'm reading this correctly, there's not yet a notion of an "output" node (that writes to a UAV or rendertarget) right? Or maybe there's a stand-in hardcoded output node that isn't in the `nodes` array we're looping over above?

Reply:
There is a notion of an output node, the lighting model node, in this case the example PBR node. The only thing that makes it unique is that it supplies a list of all of the template files, does not have any output slots/pins, and does not prepend its generated variable names with the node ID. This means that the undecorated slot names can correspond directly to expected variables referenced in the AZSL template files. Example PBR is based off of minimal PBR, modified with comment blocks for insert points and variable names that will come from the node configuration. Default code including variables with default values can be put into the template AZSL code generation block and will be replaced when the graph is compiled.

The template files themselves can be refactored to include more of that code in the node or move the bulk of the code functions that can be reused in other templates. It can be organized in many different ways.